File: ocaml/gup/state.ml
open Std
open Lwt
open Path
open CCFun

type dirty_args = {
	var: Var.t;
	path: ConcreteBase.t;
	builder_path: RelativeFrom.t;
	build: (RelativeFrom.t -> bool Lwt.t)
}

let builds_have_been_cancelled = ref false
let cancel_all_future_builds () = builds_have_been_cancelled := true

exception Version_mismatch of string
exception Invalid_dependency_file

(* lwt wrappers for with_file_in and with_file_out *)
let with_file_in path fn =
	let flags = [Unix.O_CLOEXEC; Unix.O_RDONLY] in
	Lwt_io.with_file ~flags:flags ~mode:Lwt_io.input path fn

let with_file_out ?(flags) path fn =
	let default_flags = [Unix.O_CLOEXEC; Unix.O_WRONLY; Unix.O_CREAT] in
	let flags = match flags with
		| Some f -> List.concat [default_flags; f]
		| None -> default_flags
	in
	Lwt_io.with_file ~flags:flags ~mode:Lwt_io.output path fn

type dependency_type =
	| FileDependency
	| Checksum
	| RunId
	| Builder
	| BuildTime
	| AlwaysRebuild
	| ClobbersTarget

module Log = (val Var.log_module "gup.state")
let meta_dir_name = PathComponent.name_of_string ".gup"
let deps_ext = "deps"
let new_deps_ext = "deps2"

let empty_field = "-"
let format_version = 3
let version_marker = "version: "

(* exceptionless helpers *)
let readline f =
	try%lwt
		let%lwt line = Lwt_io.read_line f in
		Lwt.return @@ Some line
	with End_of_file -> Lwt.return None

let built_targets dir =
	let contents = Sys.readdir dir in
	contents |> CCArray.filter_map (fun f ->
		if CCString.prefix ~pre:(deps_ext ^ ".") f
			then (CCString.Split.left ~by:"." f) |> CCOpt.map (PathComponent.name_of_string % snd)
			else None
	) |> Array.to_list

let resolve_builder_path ~(target:ConcreteBase.t) (path:Absolute.t) : RelativeFrom.t =
	let newbase = ConcreteBase.dirname target in
	Concrete.resolve_abs path |> ConcreteBase.of_concrete |> ConcreteBase.rebase_to newbase

class run_id id =
	object
		method is_current = id = Var.run_id
		method repr = "run_id(" ^ id ^ ")"
		method fields = [id]
		method tag = RunId
		method print out = CCFormat.(within "run_id(" ")" string out id)
	end

let current_run_id = new run_id Var.run_id

type 'a intermediate_dependencies = {
	checksum: string option ref;
	run_id: run_id option ref;
	clobbers: bool ref;
	rules: 'a list ref;
}

type dependency_class = {
	tag: dependency_type;
	num_fields: int;
}


let string_of_dependency_type = function
	| Checksum -> "checksum"
	| RunId -> "run"
	| FileDependency -> "file"
	| BuildTime -> "built"
	| Builder -> "builder"
	| AlwaysRebuild -> "always"
	| ClobbersTarget -> "clobbers"


let serializable_dependencies = [
	{ tag = FileDependency; num_fields = 3; };
	{ tag = RunId; num_fields = 1; };
	{ tag = Checksum; num_fields = 1; };
	{ tag = BuildTime; num_fields = 1; };
	{ tag = Builder; num_fields = 3; };
	{ tag = AlwaysRebuild; num_fields = 0; };
	{ tag = ClobbersTarget; num_fields = 0; };
]

let tag_assoc     = serializable_dependencies |> List.map (fun dep -> (dep.tag, dep))
let tag_assoc_str = tag_assoc |> List.map (fun (tag, dep) -> (string_of_dependency_type tag, dep))

let serializable dep = (dep#tag, dep#fields)

let write_dependency output (tag,fields) =
	let (_, typ) = List.find (fun (t, _) -> t = tag) tag_assoc in
	if List.length fields <> typ.num_fields then Error.raise_safe "invalid fields";
	Lwt_io.write_line output @@ (string_of_dependency_type tag) ^ ": " ^ (String.concat " " fields)

type dep_dirty_state = [ `clean | `dirty | `clean_after_build ]
let dirty_of_dep_state : dep_dirty_state -> bool = function
	| `dirty -> true
	| `clean | `clean_after_build -> false

let dirty_check_with_dep ~(path:RelativeFrom.t) checker args =
	let%lwt dirty = checker args in
	if dirty then Lwt.return `dirty else (
		let%lwt built = args.build path in
		if built
			then (
				Log.trace args.var (fun m->m "dirty_check_with_dep: path %s was built, rechecking" (RelativeFrom.to_string path));
				checker args |> Lwt.map (function
					| true -> `dirty
					| false -> `clean_after_build
				)
			) else Lwt.return `clean
	)

let is_mtime_mismatch ~var ~stored path =
	let%lwt current_mtime = (RelativeFrom.lift Util.get_mtime) path in
	Log.trace var PP.(fun m->m "%s: comparing stored mtime %a against current %a"
		(RelativeFrom.to_string path)
		(option Util.big_int_pp) stored
		(option Util.big_int_pp) current_mtime);
	Lwt.return @@ not @@ eq (CCOpt.compare Big_int.compare_big_int) stored current_mtime

let file_fields ~mtime ~checksum path =
	let mtime_str = CCOpt.get_or ~default:empty_field (CCOpt.map Big_int.string_of_big_int mtime) in
	let checksum_str = CCOpt.get_or ~default:empty_field checksum in
	[mtime_str; checksum_str; RelativeFrom.to_field path]

class virtual base_dependency = object (self)
	method virtual tag : dependency_type
	method virtual fields : string list
	method virtual is_dirty : dirty_args -> bool Lwt.t

	method print out =
		Format.fprintf out "<#%s: %a>"
			(string_of_dependency_type self#tag)
			PP.(list string) self#fields
end

and file_dependency ~(mtime:Big_int.big_int option) ~(checksum:string option) (path:RelativeFrom.t) =
	object (self)
		inherit base_dependency
		method tag = FileDependency
		method fields = file_fields ~mtime ~checksum path

		method private mtime = mtime
		method private is_dirty_cs checksum args =
			(* checksum-based check *)
			Log.trace args.var (fun m->m "%s: comparing using checksum %s" (RelativeFrom.to_string path) checksum);
			let resolved_path = ConcreteBase.resolve_relfrom path in
			let state = new target_state ~var:args.var resolved_path in
			let checksum_mismatch args =
				let%lwt deps = state#deps in
				let latest_checksum = CCOpt.flat_map (fun deps -> deps#checksum) deps in
				let dirty = match latest_checksum with
					| None -> true
					| Some dep_cs -> dep_cs <> checksum
				in
				Log.debug args.var PP.(fun m->m "%s: %s (stored checksum is %s, current is %a)"
					(if dirty then "DIRTY" else "CLEAN")
					(ConcreteBase.to_string resolved_path) checksum
					(option string) latest_checksum);
				Lwt.return dirty
			in
			dirty_check_with_dep ~path checksum_mismatch args |> Lwt.map dirty_of_dep_state

		method private is_dirty_mtime = dirty_check_with_dep ~path (fun args ->
			is_mtime_mismatch ~var:args.var ~stored:self#mtime path
		)

		method is_dirty args =
			let%lwt mtime_dirty = self#is_dirty_mtime args in
			match mtime_dirty with
				| `clean -> Lwt.return_false
				| `dirty | `clean_after_build -> (
					(* if mtime is unchanged after a build, try checksum anyway *)
					match checksum with
						| Some checksum -> self#is_dirty_cs checksum args
						| None -> Lwt.return (dirty_of_dep_state mtime_dirty)
				)
	end

and builder_dependency ~mtime (path:RelativeFrom.t) =
	object
		inherit base_dependency
		method tag = Builder
		method fields = file_fields ~mtime ~checksum:None path
		method is_dirty args =
			let relative_path = path |> RelativeFrom.relative in
			let builder_path = args.builder_path |> RelativeFrom.relative in
			if not (Relative.eq builder_path relative_path) then (
				Log.debug args.var (fun m->m "DIRTY: builder changed from %s -> %s"
					(Relative.to_string relative_path)
					(Relative.to_string builder_path));
				Lwt.return_true
			) else is_mtime_mismatch ~var:args.var ~stored:mtime path
	end

and always_rebuild =
	object
		inherit base_dependency
		method tag = AlwaysRebuild
		method fields = []
		method is_dirty (_:dirty_args) = Lwt.return_true
	end

and build_time time =
	object
		inherit base_dependency
		method tag = BuildTime
		method fields = [Big_int.string_of_big_int time]
		method is_dirty args =
			let path = args.path |> ConcreteBase.to_string in
			let%lwt mtime = Util.get_mtime path >>= (return $ CCOpt.get_exn) in
			Lwt.return (
				if neq Big_int.compare_big_int mtime time then (
					Log.debug args.var (fun m->m "%s mtime (%a) differs from build time (%a)" path
						Util.big_int_pp mtime
						Util.big_int_pp time);
					let log_method = ref Log.warn in
					if Util.lisdir path then
						(* dirs are modified externally for various reasons, not worth warning *)
						log_method := Log.debug
					;
					!log_method args.var (fun m->m"%s was externally modified - rebuilding" path);
					true
				)
				else false
			)
	end

and dependencies (target_path:ConcreteBase.t) (data:base_dependency intermediate_dependencies) =
	object
		method already_built = match !(data.run_id) with
			| None -> false
			| Some r -> r#is_current

		method print out =
			let open PP in
			Format.fprintf out "<#Dependencies(run=%a, cs=%a, clobbers=%b, rules=%a)>"
				(option print_obj) !(data.run_id)
				(option string) !(data.checksum)
				!(data.clobbers)
				(list print_obj) !(data.rules)

		method checksum = !(data.checksum)

		method clobbers = !(data.clobbers)

		method is_dirty ~var (buildable: Buildable.t) (build:RelativeFrom.t -> bool Lwt.t) : bool Lwt.t =
			let target_path_str = ConcreteBase.to_string target_path in
			if not (Util.lexists target_path_str) then (
				Log.debug var (fun m->m "DIRTY: %s (target does not exist)" target_path_str);
				return true
			) else (
				let script_path = Buildable.script buildable in
				let args = {
					var;
					path = target_path;
					builder_path = resolve_builder_path ~target:target_path script_path;
					build = build
				} in

				let rec iter rules =
					match rules with
					| [] ->
							Log.trace var (fun m->m "dependencies#is_dirty: %s returning false" target_path_str);
							Lwt.return_false
					| (rule::remaining_rules) ->
						Log.trace var (fun m->m "dependencies#is_dirty checking %a for path %s" print_obj rule target_path_str);
						let%lwt dirty = rule#is_dirty args in
						if dirty then (
							Log.trace var (fun m->m "dependencies#is_dirty: %s returning true" target_path_str);
							Lwt.return_true
						) else iter remaining_rules
				in
				iter !(data.rules)
			)
end

and dependency_builder ~var target_path (input:Lwt_io.input_channel) = object
	(* extracted into its own object because we can't use lwt in an object consutrctor
	 * syntax *)
	method build =
		let basedir = ConcreteBase.dirname target_path in
		let update_singleton r v =
			assert (CCOpt.is_none !r);
			r := v;
			None
		in

		let parse_line line : (dependency_class * string list) =
			let (typ, content) = CCString.Split.left line ~by:":" |> CCOpt.flat_map (fun (tag, content) ->
				CCList.find_opt (fun (prefix, _typ) -> prefix = tag) tag_assoc_str
					|> CCOpt.map (fun (_, typ) -> (typ, content))
			) |> CCOpt.get_lazy (fun () -> Error.raise_safe "invalid dep line: %s" line)
			in
			let fields = Str.bounded_split (Str.regexp " ") (CCString.ltrim content) typ.num_fields in
			(typ, fields)
		in

		let _parse input rv =
			let process_line line : base_dependency option =
				let typ, fields = parse_line (String.trim line) in
				let parse_cs cs = if cs = empty_field then None else Some cs in
				let parse_mtime mtime = if mtime = empty_field then None else Some (Big_int.big_int_of_string mtime) in
				match (typ.tag, fields) with
					| (Checksum, [cs])      -> update_singleton rv.checksum (Some cs)
					| (RunId, [r])          -> update_singleton rv.run_id (Some (new run_id r))
					| (ClobbersTarget, [])  -> (rv.clobbers := true; None)
					| (FileDependency, [mtime; cs; path]) ->
							Some (new file_dependency
								~mtime:(parse_mtime mtime)
								~checksum:(parse_cs cs)
								(RelativeFrom.of_field ~basedir path)
							)
					| (Builder, [mtime; _cs; path;]) ->
							Some (new builder_dependency
								~mtime:(parse_mtime mtime)
								(RelativeFrom.of_field ~basedir path))
					| (BuildTime, [time]) -> Some (new build_time (Big_int.big_int_of_string time))
					| (AlwaysRebuild, []) -> Some (new always_rebuild)
					| _ -> Error.raise_safe "Invalid dependency line: %s" line
			in
			let%lwt rules = Lwt_io.read_lines input
				|> Lwt_stream.filter_map process_line
				|> Lwt_stream.to_list
			in
			rv.rules := rules;
			Lwt.return rv
		in

		let%lwt (data : base_dependency intermediate_dependencies) =
			let rv = {
				checksum = ref None;
				run_id = ref None;
				clobbers = ref false;
				rules = ref [];
			} in
			let%lwt version_line = readline input in
			Log.trace var PP.(fun m->m "version_line: %a" (option string) version_line);
			let version_number = version_line |> CCOpt.flat_map (fun line ->
				if CCString.prefix ~pre:version_marker line then (
					CCString.Split.left line ~by:" " |> CCOpt.flat_map (CCInt.of_string % snd)
				) else None
			) in
			match version_number with
				| None -> raise Invalid_dependency_file
				| Some v ->
					if v <> format_version then
						raise @@ Version_mismatch ("can't read format version: " ^ (string_of_int v))
			;
			_parse input rv
		in

		Lwt.return @@ new dependencies target_path data
	end

and target_state ~var (target_path:ConcreteBase.t) =
	let base_path = ConcreteBase.dirname target_path in
	let meta_path ext =
		let target = ConcreteBase.basename target_path |> PathComponent.string_of_name_opt in
		let meta_dir = Absolute.concat (Concrete.absolute base_path) (meta_dir_name |> PathComponent.relative_of_name) in
		Absolute.concat meta_dir (Relative.of_string (ext ^ "." ^ target))
	in

	let ensure_meta_path ext =
		let p = meta_path ext in
		Absolute.lift (Util.makedirs) (Absolute.dirname p);
		p
	in

	let lock_for_ext ext = new Parallel.lock_file ~var
		~target:(ensure_meta_path ext |> Absolute.to_string)
		(meta_path (ext ^ "-lock") |> Absolute.to_string)
	in
	let deps_lock = lazy (lock_for_ext deps_ext) in
	let new_deps_lock = lazy (lock_for_ext new_deps_ext) in

	object (self)
		method meta_path ext = meta_path ext

		method path_repr = target_path |> ConcreteBase.to_string
		
		method repr =
			"TargetState(" ^ (self#path_repr) ^ ")"

		method private parse_dependencies : dependencies option Lwt.t =
			Log.trace var (fun m->m "parse_deps %s" (target_path |> ConcreteBase.to_string));
			let deps_path = self#meta_path deps_ext in
			if Absolute.lexists deps_path then (
				try%lwt
					(Lazy.force deps_lock)#use Parallel.ReadLock (fun deps_path ->
						with_file_in deps_path (fun f ->
							(new dependency_builder ~var target_path f)#build >>= (fun d -> Lwt.return (Some d))
						)
					)
				with
					| Version_mismatch _ -> (
						Log.warn var (fun m->m "Ignoring stored dependencies from incompatible version: %s"
							(Absolute.to_string deps_path));
						Lwt.return None
					)
					| Invalid_dependency_file -> (
						Log.warn var (fun m->m "Ignoring invalid stored dependencies: %s"
							(Absolute.to_string deps_path));
						Lwt.return None
					)
					| e -> (
						Log.warn var (fun m->m "Error loading %s: %s (assuming dirty)"
							(Absolute.to_string deps_path) (Printexc.to_string e));
						Lwt.return None
					)
			) else
				Lwt.return None

		method deps =
			let%lwt deps = self#parse_dependencies in
			Log.trace var PP.(fun m->m "Loaded serialized state: %a" (option print_obj) deps);
			Lwt.return deps

		method private with_dependency_lock fn : unit Lwt.t =
			(Lazy.force new_deps_lock)#use Parallel.WriteLock (fun out_filename ->
				with_file_out ~flags:[Unix.O_APPEND] out_filename fn
			)

		method private add_dependency dep : unit Lwt.t =
			self#with_dependency_lock (fun output -> write_dependency output dep)

		method private add_dependencies deps : unit Lwt.t =
			self#with_dependency_lock (fun output -> deps |> Lwt_list.iter_s (write_dependency output))

		method mark_clobbers =
			self#add_dependency (ClobbersTarget, [])

		method private file_dependency_with ~(mtime:Big_int.big_int option) ~(checksum:string option) path =
			let path = ConcreteBase.rebase_to base_path path in
			let dep = (new file_dependency ~mtime:mtime ~checksum:checksum path) in
			Log.trace var (fun m->m "Adding dependency %s -> %a"
				(ConcreteBase.basename target_path |> PathComponent.string_of_name_opt)
				print_obj dep);
			serializable dep

		method add_file_dependency_with ~(mtime:Big_int.big_int option) ~(checksum:string option) path =
			self#add_dependency (self#file_dependency_with ~mtime ~checksum path)

		method add_file_dependency path =
			let%lwt mtime = Util.get_mtime (ConcreteBase.to_string path) in
			self#add_file_dependency_with ~mtime ~checksum:None path

		method add_file_dependencies paths =
			let%lwt deps = Lwt_list.map_p (fun path ->
				let%lwt mtime = ConcreteBase.lift Util.get_mtime path in
				Lwt.return (self#file_dependency_with ~mtime ~checksum:None path)
			) paths in
			self#add_dependencies deps

		method add_checksum checksum =
			self#add_dependency (Checksum, [checksum])

		method mark_always_rebuild =
			self#add_dependency (serializable (new always_rebuild))

		method private builder_dependency path =
			let%lwt builder_mtime = (Absolute.lift Util.get_mtime) path in
			return (new builder_dependency
				~mtime:builder_mtime
				(resolve_builder_path ~target:target_path path)
			)

		method perform_build (buildable:Buildable.t) block : bool Lwt.t =
			let exe = Buildable.script buildable in
			assert (Sys.file_exists (Absolute.to_string exe));
			Log.trace var (fun m->m "perform_build %s" (ConcreteBase.to_string target_path));

			(* TODO: quicker mtime-based check *)
			let still_needs_build = (fun deps ->
				Log.trace var (fun m->m "Checking if %s still needs build after releasing lock"
					(ConcreteBase.to_string target_path));
				match deps with
					| Some d -> not d#already_built
					| None -> true
			) in

			let build = (fun deps_path ->
				if (!builds_have_been_cancelled) then raise Error.BuildCancelled;
				let%lwt deps = self#deps in
				if still_needs_build deps then (
					let%lwt builder_dep = self#builder_dependency exe in
					let temp = ensure_meta_path new_deps_ext |> Absolute.to_string in
					with_file_out temp (fun file ->
						Lwt_io.write_line file (version_marker ^ (string_of_int format_version)) >>= fun () ->
						(* TODO: make Dependencies module to store init stuff *)
						write_dependency file (serializable builder_dep) >>= fun () ->
						write_dependency file (serializable current_run_id)
					) >>= fun () ->
					let%lwt built = try%lwt
						block deps
					with ex ->
						let%lwt () = Lwt_unix.unlink temp in
						raise ex
					in
					let%lwt () = if built then (
						let%lwt mtime = (ConcreteBase.lift Util.get_mtime) target_path in
						mtime |> Lwt_option.may (fun time ->
							with_file_out ~flags:[Unix.O_APPEND] temp (fun output ->
								let timedep = new build_time time in
								write_dependency output (serializable timedep)
							)
						) >>= fun () ->
						Lwt_unix.rename temp deps_path
					) else return_unit in
					return built
				) else Lwt.return false
			) in

			let%lwt built = (Lazy.force deps_lock)#use Parallel.WriteLock build in
			return built
	end

let of_buildable ~var b = new target_state ~var (Buildable.target_path b)

File: python/gup/state.py
import os
import logging
import errno

from .util import *
from .log import getLogger
from .gupfile import Builder
from .parallel import Lock
from .var import RUN_ID
_log = getLogger(__name__)

META_DIR = '.gup'

class VersionMismatch(ValueError): pass

class _dirty_args(object):
	def __init__(self, deps, base, builder_path, build_dependency):
		self.deps = deps
		self.base = base
		self.builder_path = builder_path
		self.build_dependency = build_dependency

def dirty_check_with_dep(path, check_fn, args):
	dirty = check_fn()
	if dirty: return True
	built = args.build_dependency(path)
	if built:
		_log.trace("dirty_check_with_dep: path %s was built, rechecking", path);
		return check_fn()
	else:
		return False

class TargetState(object):
	_dep_lock = None

	def __init__(self, p):
		self.path = p
	
	def __repr__(self):
		return 'TargetState(%r)' % (self.path,)
	
	@staticmethod
	def built_targets(dir):
		'''
		Returns the target names which have metadata stored in `dir`
		'''
		return [f[:-5] for f in os.listdir(dir) if f.endswith('.deps')]

	def meta_path(self, ext):
		base, target = os.path.split(self.path)
		meta_dir = os.path.join(base, META_DIR)
		return os.path.join(meta_dir, "%s.%s" % (target, ext))

	def _ensure_meta_path(self, ext):
		p = self.meta_path(ext)
		mkdirp(os.path.dirname(p))
		return p
	
	def _ensure_dep_lock(self):
		if not self._dep_lock:
			self._dep_lock = Lock(self._ensure_meta_path('deps.lock'))
		return self._dep_lock

	def deps(self):
		rv = None
		deps_path = self.meta_path('deps')
		if not os.path.exists(deps_path):
			_log.trace("Not loading missing deps at %s", deps_path)
			return rv

		with self._ensure_dep_lock().read():
			try:
				f = open(deps_path)
			except IOError as e:
				if e.errno != errno.ENOENT: raise
			else:
				try:
					with f:
						rv = Dependencies(self.path, f)
				except VersionMismatch as e:
					_log.debug("Ignoring stored dependencies from incompatible version: %s", deps_path)
				except Exception as e:
					_log.debug("Error loading %s: %s (assuming dirty)", deps_path, e)
		_log.trace("Loaded serialized state from %s: %r" % (deps_path, rv,))
		return rv

	def create_lock(self):
		if self.lockfile is None:
			self.lockfile = Lock(self.meta_path('lock'))

	def add_dependency(self, dep):
		lock = Lock(self.meta_path('deps2.lock'))
		_log.debug('add dep: %s -> %s' % (self.path, dep))
		with lock.write():
			with open(self.meta_path('deps2'), 'a') as f:
				dep.append_to(f)
	
	def mark_clobbers(self):
		self.add_dependency(ClobbersTarget())
	
	def perform_build(self, builder, do_build):
		exe = builder.path
		assert os.path.exists(exe)
		def still_needs_build(deps):
			_log.trace("checking if %s still needs build after releasing lock" % self.path)
			return deps is None or (not deps.already_built())

		with self._ensure_dep_lock().write():
			deps = self.deps()
			if not still_needs_build(deps):
				return False

			builder_dep = BuilderDependency.relative_to_target(self.path,
				path=builder.realpath,
				mtime=get_mtime(builder.realpath))

			_log.trace("created dep %s from builder %r" % (builder_dep, exe))
			temp = self._ensure_meta_path('deps2')
			with open(temp, 'w') as f:
				Dependencies.init_file(f)
				builder_dep.append_to(f)
			try:
				built = do_build(deps)
			except:
				os.remove(temp)
				raise
			else:
				if built:
					# always track the build time
					built_time = get_mtime(self.path)
					if built_time is not None:
						with open(temp, 'a') as f:
							BuildTime(built_time).append_to(f)
					rename(temp, self.meta_path('deps'))
				return built

class Dependencies(object):
	FORMAT_VERSION = 3
	def __init__(self, path, file):
		self.path = path
		self.rules = []
		self.checksum = None
		self.clobbers = False
		self.runid = None

		if file is None:
			self.rules.append(NeverBuilt())
		else:
			version_line = file.readline().strip()
			_log.trace("version_line: %s" % (version_line,))
			if not version_line.startswith('version:'): raise ValueError("Invalid file")
			_, file_version = version_line.split(' ')
			if int(file_version) != self.FORMAT_VERSION:
				raise VersionMismatch("can't read format version %s" % (file_version,))

			while True:
				line = file.readline()
				if not line: break
				dep = Dependency.parse(line.rstrip())
				if isinstance(dep, Checksum):
					assert self.checksum is None
					self.checksum = dep.value
				elif isinstance(dep, RunId):
					assert self.runid is None
					self.runid = dep
				elif isinstance(dep, ClobbersTarget):
					self.clobbers = True
				else:
					self.rules.append(dep)
	
	def is_dirty(self, builder, build_dependency):
		assert isinstance(builder, Builder)
		if not os.path.lexists(self.path):
			_log.debug("DIRTY: %s (target does not exist)", self.path)
			return True

		base = os.path.dirname(self.path)
		builder_path = os.path.relpath(builder.realpath, base)

		dirty_args = _dirty_args(deps=self, base=base, builder_path=builder_path, build_dependency=build_dependency)
		for rule in self.rules:
			d = rule.is_dirty(dirty_args)
			if d:
				_log.trace('DIRTY: %s (from rule %r)', self.path, rule)
				return True
		_log.trace('is_dirty: %s returning %r', self.path, False)
		return False
	
	def already_built(self):
		return self.runid.is_current()

	@classmethod
	def init_file(cls, f):
		f.write('version: %s\n' % (cls.FORMAT_VERSION,))
		RunId.current().append_to(f)
	
	def __repr__(self):
		return 'Dependencies<runid=%r, checksum=%s, %r>' % (self.runid, self.checksum, self.rules)

class Dependency(object):
	recursive = False
	@staticmethod
	def parse(line):
		_log.trace("parsing line: %s" % (line,))
		for candidate in [
				FileDependency,
				BuilderDependency,
				AlwaysRebuild,
				Checksum,
				RunId,
				ClobbersTarget,
				BuildTime]:
			if line.startswith(candidate.tag):
				cls = candidate
				break
		else:
			raise ValueError("unknown dependency line: %r" % (line,))
		fields = line.split(' ', cls.num_fields)[1:]
		return getattr(cls, 'deserialize', cls)(*fields)

	def append_to(self, file):
		line = self.tag + ' ' + ' '.join(self.fields)
		assert "\n" not in line
		file.write(line + "\n")
	
	def __repr__(self):
		return '%s(%s)' % (type(self).__name__, ', '.join(map(repr, self.fields)))

class NeverBuilt(object):
	fields = []
	def is_dirty(self, args):
		_log.debug('DIRTY: never built')
		return True
	def append_to(self, file): pass

class AlwaysRebuild(Dependency):
	tag = 'always:'
	num_fields = 0
	fields = []
	def is_dirty(self, _):
		_log.debug('DIRTY: always rebuild')
		return True

class FileDependency(Dependency):
	num_fields = 3
	tag = 'file:'
	recursive = True

	def __init__(self, mtime, checksum, path):
		self.path = path
		self.checksum = checksum
		self.mtime = mtime
	
	@classmethod
	def relative_to(cls, rel_root, mtime, path):
		rel_path = os.path.relpath(path, rel_root)
		return cls(mtime=mtime, checksum=None, path=rel_path)

	@classmethod
	def relative_to_target(cls, target, mtime, path):
		return cls.relative_to(os.path.dirname(target), mtime=mtime, path=path)
	
	@classmethod
	def of_target(cls, parent_target, target, mtime):
		rv = cls.relative_to_target(parent_target, mtime=mtime, path=target.path)
		deps = target.state.deps()
		if deps and deps.checksum:
			rv.checksum = deps.checksum
		return rv
	
	@classmethod
	def deserialize(cls, mtime, checksum, path):
		return cls(
			None if mtime == '-' else int(mtime),
			None if checksum == '-' else checksum,
			path)
	
	@property
	def fields(self):
		return [
			'-' if self.mtime is None else str(self.mtime),
			self.checksum or '-',
			self.path]

	def full_path(self, base):
		if os.path.isabs(self.path): return self.path
		return os.path.normpath(os.path.join(base, self.path))

	def is_dirty(self, args):
		base = args.base
		path = self.full_path(base)
		self._target = path

		def mtime_dirty():
			current_mtime = get_mtime(path)
			if current_mtime != self.mtime:
				_log.debug("DIRTY: %s (stored mtime is %r, current is %r)" % (self.path, self.mtime, current_mtime))
				return True
			return False
		
		dirty = dirty_check_with_dep(path, mtime_dirty, args)

		if dirty and self.checksum is not None:
			def checksum_dirty():
				_log.trace("%s: comparing using checksum %s", self.path, self.checksum)
				state = TargetState(path)
				deps = state.deps()
				checksum = deps and deps.checksum
				if checksum != self.checksum:
					_log.debug("DIRTY: %s (stored checksum is %s, current is %s)", self.path, self.checksum, checksum)
					return True
				return False

			return dirty_check_with_dep(path, checksum_dirty, args)
		else:
			return dirty

class BuilderDependency(FileDependency):
	tag = 'builder:'
	recursive = False

	def is_dirty(self, args):
		builder_path = args.builder_path

		assert not os.path.isabs(builder_path)
		assert not os.path.isabs(self.path)
		if builder_path != self.path:
			_log.debug("DIRTY: builder changed from %s -> %s" % (self.path, builder_path))
			return True
		return super(BuilderDependency, self).is_dirty(args)

class Checksum(Dependency):
	tag = 'checksum:'
	num_fields = 1

	def __init__(self, cs):
		self.value = cs
		self.fields = [cs]
	
	@staticmethod
	def _add_file(sh, f):
		if sh is None:
			import hashlib
			sh = hashlib.sha1()
		while True:
			b = f.read(4096)
			if not b: break
			sh.update(b)
		return sh

	@classmethod
	def from_stream(cls, f):
		sh = cls._add_file(None, f)
		return cls(sh.hexdigest())
	
	@classmethod
	def from_files(cls, filenames):
		sh = None
		for filename in filenames:
			with open(filename, 'rb') as f:
				sh = cls._add_file(sh, f)
		return cls(sh.hexdigest())

class BuildTime(Dependency):
	tag = 'built:'
	num_fields = 1

	def __init__(self, mtime):
		assert mtime is not None
		self.value = mtime
		self.fields = [str(mtime)]

	@classmethod
	def deserialize(cls, mtime):
		return cls(int(mtime))
	
	def is_dirty(self, args):
		path = args.deps.path

		mtime = get_mtime(path)
		assert mtime is not None
		_log.debug("comparing stored mtime %s to %s", self.value, mtime)
		if mtime != self.value:
			log_method = _log.warn
			if os.path.isdir(path):
				# dirs are modified externally for various reasons, not worth warning
				log_method = _log.debug
			log_method("%s was externally modified - rebuilding" % (path,))
			return True
		return False

class RunId(Dependency):
	tag = 'run:'
	num_fields = 1

	def __init__(self, runid):
		self.value = runid
		self.fields = [runid]

	@classmethod
	def current(cls):
		return cls(RUN_ID)

	def is_current(self):
		return self.value == RUN_ID

class ClobbersTarget(Dependency):
	tag = 'clobbers:'
	num_fields = 0
	fields = []
